Issue/3110 modules v2 moduletool commands

sanderr opened the pull request:
# Description

Reviewed all moduletool commands for v2 compatibility.

closes #3110

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] ~~Sufficient test cases (reproduces the bug/tests the requested feature)~~
- [x] Correct, in line with design
- [x] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

arnaudsjs commented on src/inmanta/module.py:
```diff
@@ -2543,6 +2496,58 @@ def add_module_requirement_persistent(self, requirement: InmantaModuleRequiremen
+
+    def push(self) -> None:
+        """
+        Run a git status on this module
```
```suggestion
        Run a git push on this module
```

arnaudsjs commented on src/inmanta/moduletool.py:
```diff
@@ -528,7 +531,20 @@ def get_modules(self, module: Optional[str] = None) -> List[Module]:
+                raise Exception(f"Directory {module_dir} already exists")
+            cookiecutter(
+                "https://github.com/inmanta/inmanta-module-template.git",
+                checkout="v2",
```
Shouldn't we perform a checkout from master? The intention is to have the V2 template on the master branch after the release, no? It looks like no `v2` branch exist at the moment. Was this functionality tested?

wouterdb replied:
I expect we will need the V1 a little longer until we have set up the proper distribution channels for v2 modules.

sanderr replied:
I'd discussed both this and https://github.com/inmanta/inmanta-core/pull/3473#discussion_r752924840 with Wouter yesterday but I'd forgotten to share the conclusions. The main reason we need to keep v1 on master for now is that new v2 modules require a v2 std, which we haven't decided how to distribute yet. The `v2` branch does not exist yet but this is part of inmanta/inmanta-module-template#13.

arnaudsjs commented on src/inmanta/moduletool.py:
```diff
@@ -584,42 +603,54 @@ def list(self, requires: bool = False) -> None:
+                matches = version == reqv
+                editable = True
+            else:
+                reqv = ",".join(req.version_spec_str() for req in specs[name] if req.specs) or "*"
```
The usage of `reqv` is inconsistent between `V1` and `V2` modules.

* For V1 modules, `reqv` contains a list of version that match the version spec
* For V2 modules, `reqv` contains the version specs.

Isn't this strange?

sanderr replied:
I agree somewhat strange, though both Wouter and I are of the opinion that the v2 format is an improvement. I opted to stick with the stricter approach for v1 because it needs to take the compiler version into account, making it more complex than just the version specs. We could clean that up as well but I'm not sure it's worth the effort for v1.
A small remark though, for v1 modules, if I understand correctly, it never contains a list, always a single version: the latest version with a compatible compiler version constraint.

arnaudsjs commented on src/inmanta/moduletool.py:
```diff
@@ -584,42 +603,54 @@ def list(self, requires: bool = False) -> None:
             version_length = max(len(version), len(reqv), version_length)
 
-            table.append((name, version, reqv, version == reqv))
+            table.append((name, generation, editable, version, reqv, matches))
```
The `version_length` variable is not used anywhere. We can clean that up in my opinion.

sanderr replied:
Done.

arnaudsjs reviewed (changes requested):
See review comments.

wouterdb reviewed (changes requested):
Why no testcases?

sanderr commented on changelogs/unreleased/3110-modules-v2-moduletool-commands.yml:
```diff
@@ -0,0 +1,6 @@
+description: Reviewed all moduletool commands for v2 compatibility
+change-type: major
+sections:
+  deprecation-note: "The `inmanta module list -r` command has been deprecated"
```
Done

wouterdb reviewed (commented):
testcases! (I comment so you can re-request when you are done)

sanderr commented:
Also from the demo meeting:
- check `matches` behavior for release mode master has not been changed by this PR
- mention deprecation in `inmanta module list --help` output
- Generation -> module type
- 1/0 -> yes/no

sanderr commented:
Re-requesting review because of fundamental changes:
- Dropped `--force-reinstall` from module install command as discussed on slack. Modified test cases to be compatible.
- Extended `Project.requires` to include v2 requirements from `requirements.txt`

Non-fundamental changes:
- added tests
- suggestions from demo (see [my comment above](https://github.com/inmanta/inmanta-core/pull/3473#issuecomment-974116096))

arnaudsjs commented on src/inmanta/moduletool.py:
```diff
@@ -528,7 +537,21 @@ def get_modules(self, module: Optional[str] = None) -> List[Module]:
             return self.get_project(load=True).sorted_modules()
 
-    def create(self, name: str) -> None:
+    def create(self, name: str, v1: bool, no_input: bool = False) -> None:
```
We should add a comment to explain the the `no_input` argument is only relevant for V2 modules. Or we should change the name of the argument.

sanderr replied:
Added a docstring, inlcuding this.

wouterdb commented on tests/moduletool/test_install.py:
```diff
@@ -753,3 +764,101 @@ def test_project_install_with_install_mode(
+    )
+
+    cwd = os.getcwd()
+    try:
```
Why is this?

sanderr replied:
This was meant to wrap the whole block below, but apparently it's not required because `snippetcompiler` does `Project.set`. I'll remove it.

inmantaci commented:
Processing this pull request

inmantaci commented:
Merged into branches master in b27d17080f3fa6a26fca4afd3755fd8d6b89a728
